zap/src/output/luau/mod.rs

```diff
@@ -18,69 +15,526 @@ pub trait Output {
-
-			Stmt::End => self.push_line("end"),
-		};
-
-		if matches!(
-			stmt,
-			Stmt::NumFor { .. } | Stmt::GenFor { .. } | Stmt::If(..) | Stmt::ElseIf(..) | Stmt::Else
-		) {
-			self.indent();
-		};
+			Ty::Struct(struct_ty) => {
+				for (field_name, field_ty) in &struct_ty.fields {
+					self.push_ser(&format!("{from}.{field_name}"), field_ty, checks);
+				}
+			}
+
+			Ty::Instance(class) => {
+				if checks && class.is_some() {
+					self.push_line(&format!("assert({from}:IsA(\"{}\"))", class.unwrap()));
+				}
+
+				self.push_line(&format!("table.insert(outgoing_inst, {from})"));
+			}
+
+			Ty::Color3 => {
+				self.push_line("alloc(3)");
+				self.push_line(&format!("buffer.writeu8(outgoing_buff, outgoing_apos, {from}.R * 255)"));
+				self.push_line(&format!(
+					"buffer.writeu8(outgoing_buff, outgoing_apos + 1, {from}.G * 255)"
+				));
+				self.push_line(&format!(
+					"buffer.writeu8(outgoing_buff, outgoing_apos + 1, {from}.B * 255)"
+				));
+			}
+
+			Ty::Vector3 => {
+				self.push_line("alloc(12)");
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos, {from}.X)"));
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos + 4, {from}.Y)"));
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos + 8, {from}.Z)"));
+			}
+
+			Ty::AlignedCFrame => {
+				self.push_line(&format!(
+					"local axis_alignment = table.find(CFrameSpecialCases, {from}.Rotation)"
+				));
+				self.push_line("assert(axis_alignment)");
+				self.push_line("alloc(1)");
+				self.push_line("buffer.writeu8(outgoing_buff, outgoing_apos, axis_alignment)");
+
+				self.push_line("alloc(12)");
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos, {from}.Position.X)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 4, {from}.Position.Y)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 8, {from}.Position.Z)"
+				));
+			}
+
+			Ty::CFrame => {
+				self.push_line(&format!("local axis, angle = {from}:ToAxisAngle()"));
+				self.push_line("axis = axis * angle");
+
+				self.push_line("alloc(12)");
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos, {from}.Position.X)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 4, {from}.Position.Y)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 8, {from}.Position.Z)"
+				));
+
+				self.push_line("alloc(12)");
+				self.push_line("buffer.writef32(outgoing_buff, outgoing_apos, axis.X)");
+				self.push_line("buffer.writef32(outgoing_buff, outgoing_apos + 4, axis.Y)");
+				self.push_line("buffer.writef32(outgoing_buff, outgoing_apos + 8, axis.Z)");
+			}
+
+			Ty::Boolean => {
+				self.push_line("alloc(1)");
+				self.push_line(&format!(
+					"buffer.writeu8(outgoing_buff, outgoing_apos, {from} and 1 or 0)"
+				))
```

Comment:
micro optimisation - `if {from} then 1 else 0` might be faster than `{from} and 1 or 0`

Reply:
I think luau compiles it down either way, but it's more clear as an if expression